src/NUnitEngine/nunit.engine.core/Services/ExtensionService.cs

```diff
@@ -325,35 +338,57 @@ private int ProcessAddinsFiles(IDirectory startDir, bool fromWildCard)
         /// lines and comments started by # are ignored.
         /// </summary>
-        private void ProcessAddinsFile(IDirectory baseDir, string fileName, bool fromWildCard)
+        private void ProcessAddinsFile(IDirectory baseDir, IFile addinsFile, bool fromWildCard)
         {
-            log.Info("Processing file " + fileName);
+            log.Info("Processing file " + addinsFile.FullName);
 
-            using (var rdr = new StreamReader(fileName))
+            foreach (var entry in _addinsReader.Read(addinsFile))
             {
-                while (!rdr.EndOfStream)
+                bool isWild = fromWildCard || entry.Contains("*");
+                var args = GetBaseDirAndPattern(baseDir, entry);
+                if (entry.EndsWith("/"))
                 {
-                    var line = rdr.ReadLine();
-                    if (line == null)
-                        break;
-
-                    line = line.Split(new char[] { '#' })[0].Trim();
-
-                    if (line == string.Empty)
-                        continue;
-
-                    if (Path.DirectorySeparatorChar == '\\')
-                        line = line.Replace(Path.DirectorySeparatorChar, '/');
+                    foreach (var dir in _directoryFinder.GetDirectories(args.Item1, args.Item2))
+                    {
+                        ProcessDirectory(dir, isWild);
+                    }
+                }
+                else
+                {
+                    foreach (var file in _directoryFinder.GetFiles(args.Item1, args.Item2))
+                    {
+                        ProcessCandidateAssembly(file.FullName, isWild);
+                    }
+                }
+            }
+        }
 
-                    bool isWild = fromWildCard || line.Contains("*");
-                    if (line.EndsWith("/"))
-                        foreach (var dir in _directoryFinder.GetDirectories(baseDir, line))
-                            ProcessDirectory(dir, isWild);
-                    else
-                        foreach (var file in _directoryFinder.GetFiles(baseDir, line))
-                            ProcessCandidateAssembly(file.FullName, isWild);
+        private Backports.Tuple<IDirectory, string> GetBaseDirAndPattern(IDirectory baseDir, string path)
+        {
+            if (Path.IsPathFullyQualified(path))
+            {
+                if (path.EndsWith("/"))
+                {
+                    return new Backports.Tuple<IDirectory, string>(_fileSystem.GetDirectory(path), string.Empty);
+                }
+                else
+                {
+                    return new Backports.Tuple<IDirectory, string>(_fileSystem.GetDirectory(System.IO.Path.GetDirectoryName(path)), System.IO.Path.GetFileName(path));
                 }
             }
+            else if (!IsPathRelative(path))
+            {
+                throw new NUnitEngineException($"The path '{path}' is not relative.");
+            }
```

Comment:
I'm a little confused by this clause sorry. This should never be hit, right - a path that's neither fully qualified nor relative? 

An exception thrown here could be displayed to the end user, so if this being thrown would be an indication of an error in the code, how about we go with something more descriptive? Maybe something like "Error when discovering extensions for path: {path}. Unexpected path format."

Reply:
That exception is thrown when an absolute Unix path is found on a Windows system and vice versa. The exception is intended to help the user isolate the cause of the error.
I had the following scenarios in my mind when I added that additional check:

- A user copies an addins-file with an absolute path from a Windows to a Unix machine. The expected addins would not load because the absolute path is treated as a relative path and will probably not be resolved.
- A user copies an addins-file with an absolute path from a Unix to a Windows machine. The expected addins would not load because the absolute path is treated as a relative path on the current working drive.

I absolutely agree that the exception-message is not useful in this case. It would be better to explain what to do to solve the error.
On the other hand, maybe I should remove the check because it is not really helpful... 🤔

Reply:
That makes sense. 🙂 

I don't think I have a particularly strong opinion on whether we skip or throw here. I guess one thought is that any such path would get caught as a Missing Directory slightly later on anyway (I think?), which potentially means we're not gaining a lot from this extra check?

Happy to go with whatever you think is best here - you have a better understanding of the details of this part of the codebase than I do! My only thinking is that if we do keep the exception, I think it would be good for it to have a more specific error message.

Reply:
Unfortunately there is no message or exception thrown when an `.addins`-file contains a reference to a non-existing item.

But to always handle that "unresolvable reference" scenario in the same way, I removed the check and the exception. This way we achieve a better readability for developers and users don't wonder why sometimes an exception is thrown for an invalid entry and sometimes not.